### PR TITLE
User log in and new user options

### PR DESCRIPTION
**Pull request by JAAR91**
This feature includes new user registration and user login, this works together with post and comments, allowing to post only registered users and the same with comments, also modify and delete a post checks if the user is logged in or not.

### CODE REVIEW

File: app/controllers/post_controller.rb [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+  end
+
+  def new
+    @post = Post.new
+  end
+
+  def create
+    @post = Post.new(post_params)
+    p @post
+    p @post.save
```
Comment (rukundoeric) [inline app/controllers/post_controller.rb]: debugging code should be removed when there are not needed anymore.

File: app/views/devise/passwords/edit.html.erb [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+<%= form_for(resource, as: resource_name, url: password_path(resource_name), html: { method: :put }) do |f| %>
+  <%= render "devise/shared/error_messages", resource: resource %>
+  <%= f.hidden_field :reset_password_token %>
+
+  <div class="field">
+    <%= f.label :password, "New password" %><br />
+    <% if @minimum_password_length %>
+      <em>(<%= @minimum_password_length %> characters minimum)</em><br />
+    <% end %>
```
Comment (rukundoeric) [inline app/views/devise/passwords/edit.html.erb:11]: views should be logic-free. you should not have any logic in the views, specifically `if-else` statements.

File: app/views/devise/mailer/email_changed.html.erb [hunk 3]
```diff
@@ -0,0 +1,7 @@
+<p>Hello <%= @email %>!</p>
+
+<% if @resource.try(:unconfirmed_email?) %>
+  <p>We're contacting you to notify you that your email is being changed to <%= @resource.unconfirmed_email %>.</p>
+<% else %>
+  <p>We're contacting you to notify you that your email has been changed to <%= @resource.email %>.</p>
+<% end %>
```
Comment (rukundoeric) [inline app/views/devise/mailer/email_changed.html.erb:7]: views should be logic-free. you should not have any logic in the views, specifically `if-else` statements.

File: app/views/comment/_comment.html.erb [hunk 4]
```diff
@@ -0,0 +1,18 @@
+
+<% post.order('created_at')%>
+<% post.order('created_at DESC').each do |comment| %>
+  <div class="container bg-light my-2 py-2 border">
+    <div class="d-flex">
+      <p class="text-black-50"><%= comment.user.email %></p>
+      <p class="mx-5 text-black-50"><%= comment.created_at %></p>
+    </div>
+    <p><%= comment.info %></p>
+    <% if current_user.id == comment.user.id %>
```
Comment (rukundoeric) [inline app/views/comment/_comment.html.erb]: views should be logic-free. you should not have any logic in the views, specifically `if-else` statements.

File: app/views/devise/registrations/edit.html.erb [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+  <%= render "devise/shared/error_messages", resource: resource %>
+
+  <div class="field">
+    <%= f.label :email %><br />
+    <%= f.email_field :email, autofocus: true, autocomplete: "email" %>
+  </div>
+
+  <% if devise_mapping.confirmable? && resource.pending_reconfirmation? %>
+    <div>Currently waiting confirmation for: <%= resource.unconfirmed_email %></div>
+  <% end %>
```
Comment (rukundoeric) [inline app/views/devise/registrations/edit.html.erb:13]: views should be logic-free. you should not have any logic in the views, specifically `if-else` statements.

File: app/views/devise/registrations/edit.html.erb [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+  <div class="field">
+    <%= f.label :password %> <i>(leave blank if you don't want to change it)</i><br />
+    <%= f.password_field :password, autocomplete: "new-password" %>
+    <% if @minimum_password_length %>
+      <br />
+      <em><%= @minimum_password_length %> characters minimum</em>
+    <% end %>
```
Comment (rukundoeric) [inline app/views/devise/registrations/edit.html.erb:21]: views should be logic-free. you should not have any logic in the views, specifically `if-else` statements.

File: app/views/post/show.html.erb [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+                    <p class="shadown ms-auto text-end border-bottom py-2">
+                    by <%= @post.user.email %>
+                    </p>
+                <%end%>
+                <% if current_user.id == @post.user_id %>
+                    <div class="d-flex flex-row justify-content-end">
+                        <%= link_to "Edit", edit_user_post_path(current_user.id, @post.id), class:'mx-2 text-decoration-none link-primary' %></li>
+                        <%= link_to "Destroy", user_post_path(current_user.id,@post.id), method: :delete, data: { confirm: "Are you sure?" }, class:'link-danger mx-2 text-decoration-none' %>
+                    </div>
+                <%end%>
```
Comment (rukundoeric) [inline app/views/post/show.html.erb]: views should be logic-free. you should not have any logic in the views, specifically `if-else` statements.

File: app/views/post/show.html.erb [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+                
+            </div>
+            <div class="border border-3 rounded shadown p-3 my-3">
+                <%= render 'comment/form' %>
+            </div>
+            <div class="border border-3 rounded shadown p-3">
+                <p class="w-100 border-bottom my-3 text-black-50 fs-5">Comments</p>
+                <% if @post.comment.length.zero? %>
+                <p class="text-black-50 text-center">No comments yet</p>
```
Comment (rukundoeric) [inline app/views/post/show.html.erb]: views should be logic-free. you should not have any logic in the views, specifically `if-else` statements.

File: app/views/shared/_navbar.html.erb [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+                Not logged in
+            <%end%>  
+          </a>
+          <ul class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
+             <% if user_signed_in? %>
+                <li><%= link_to 'Sign Out', destroy_user_session_path, method: :delete, class:'nav-link' %></li>
+             <%else%>
+                <li><%= link_to 'Register', new_user_registration_path, class:'nav-link' %></li>
+                <li><%= link_to 'Sign In', new_user_session_path, class:'nav-link' %></li>
+             <%end%>
```
Comment (rukundoeric) [inline app/views/shared/_navbar.html.erb]: views should be logic-free. you should not have any logic in the views, specifically `if-else` statements.

File: config/routes.rb [hunk 10]
```diff
@@ -0,0 +1,12 @@
+Rails.application.routes.draw do
+  devise_for :users
+  root to: 'post#index'
+
+  resources :user do
+    resources :post, only: [:new, :create, :show, :edit, :destroy, :update]
```
Comment (rukundoeric) [inline config/routes.rb]: [OPTIONAL] There is no need to use `only:` keyword when you are going to use all methods from resources.

**Review by rukundoeric — CHANGES_REQUESTED** [review]
# Status: ♻️ CHANGES REQUIRED ♻️
_Great job guys so far 👏 💯 👍. However, there are few issues we need to address before we move on, please check the comments below._

- ### To Highlight
  -  Follows gitwork-flow. 💯
  -  App is functional.  💯 

- ### To Improve
  - [x] [MANDATORY] debugging code should be removed when there are not needed anymore.
  - [x] [MANDATORY] views should be logic-free. you should not have any logic in the views, specifically `if-else` statements.
  - [x] [OPTIONAL] There is no need to use the` only:` keyword when you are going to use all methods from resources.
  
_Please make all the required changes before you submit another code review. If you have any problems or concerns, you can reach out to me on Slack: Rukundo Eric 🚀 . I wish you all the very best 🥇_

### _Keep working hard._
### _Happy Coding😃_
-----------------------------------------------------------------------------------------------------------------------------------------
**Regards,
[Rukundo Eric](https://rukundo-eric.herokuapp.com/) ,
[LinkedIn](https://www.linkedin.com/in/rukundo-eric/), 
Slack: Rukundo Eric**

**Comment by rukundoeric** [comment]
Check out [This](https://github.com/rukundoeric/book-u/blob/develop/app/helpers/application_helper.rb) in case you want to understand how you are going to migrate your logic into helpers.

**Review by GzimAsani — APPROVED** [review]
# STATUS: ACCEPTED :green_circle: :partying_face: 

## Hi @JAAR91 thank you for making all the required changes, you project now is completed and you may continue with the next one :+1: 

_**PS: Don't forget to submit the project completion form**_

# HAPPY CODING :computer: :desktop_computer:

**Pull request merged**